### ISSUE TITLE
Reliable Mode with multiple consumers

### ISSUE THREAD

**Issue by CodeJjang** [capriza/node-busmq#40]
> Note: This mode does not work well with multiple consumers. The behavior of multiple clients consuming in reliable mode from the same queue is undefined.

Reliable mode is required when you actually expect your app to be stable.
It's a pity that you cannot spawn multiple consumers in that mode, because that's the whole idea of the queues mechanism.. to be able to scale.

Is this behavior is going to change soon?

**Comment by fujifish** [comment]
The reason is that these are not _job_ queues, they are _message_ queues where **order is guaranteed**.

Imagine two consumers consuming in reliable mode. There are 3 messages in the queue, consumer A consumes message 1 and consumer B consumes message 2. Now imagine that consumer A crashes and did not have a chance to handle the message, but consumer B did and performed an 'ACK' on message 2. Because this is a message queue, acking message 2 also implicitly acks message 1 (as also happens in TCP). You end up **losing message 1** and the next message you receive from the queue will be message 3.

If what you need is a reliable _job queue_ where order is not important, take a look  [Kue](https://github.com/learnboost/kue), [Bull](https://github.com/OptimalBits/bull) or [Convoy](https://github.com/gosquared/convoy)

You can also look at [Disque](https://github.com/antirez/disque) from the creator of redis.

> Disque queues only provides best effort ordering. Each queue sorts messages based on the job creation time, which is obtained using the wall clock of the local node where the message was created (plus an incremental counter for messages created in the same millisecond), so messages created in the same node are normally delivered in the same order they were created. This is not causal ordering since correct ordering is violated in different cases: when messages are re-issued because they are not acknowledged, because of nodes local clock drifts, and when messages are moved to other nodes for load balancing and federation (in this case you end with queues having jobs originated in different nodes with different wall clocks). However all this also means that normally messages are not delivered in random order and usually messages created first are delivered first.
> 
> Note that since Disque does not provide strict FIFO semantics, technically speaking it should not be called a message queue, and it could better identified as a message broker. However I believe that at this point in the IT industry a message queue is often more lightly used to identify a generic broker that may or may not be able to guarantee order in all cases. Given that we document the semantics very clearly, I grant myself the right to call Disque a message queue anyway.

**Comment by tlenex** [comment]
I think that simple solution to this is to create separate queues for each consumer, but additionally they will need to handle duplication your data if it's important for you + performance may drop.

For tasks like you mentioned @CodeJjang I recommend using [apache kafka](http://kafka.apache.org/) - it's more like a event stream, which offers both, order and scalability. It's very stable too. There are some node modules for it.
Kafka is setting up an `zookeeper`, which remembers where your `consumerId` has ended consuming messages. Maybe this is the way to go for `busmq` module? :)

**Comment by CodeJjang** [comment]
@fujifish:
Then how do you ensure the order without acks?
Let's assume unreliable mode, consumer A takes message 1, consumer B takes message 2.
Now, if consumer B logic is quicker than consumer A, consumer B can finish it's execution before consumer A (consumer A could take much time or even crash), how would the fact that the order of insertion was preserved helped here, if there's no guarantee that consumers finish execution in particular order?

@tlenex:
I have already made a different queue for each consumer, and I dispatch different jobs (of different "job types") for every queue, but then i realized I couldn't work reliably and scale consumers (on a particular queue).

So I have question to both of you:

Which of the alternatives to busmq seems the most suitable one?

Basically I'd like to have different kind of jobs, some can be parallelized and some need to be sequential (such that one consumer is the producer of another job, etc).

I'd also like to be able to perform the jobs reliably (so if one consumer fails handling it, it would get back to the queue), be able to scale up consumers on a certain queue and most importantly that each job would be done only once of course...

Is any of the suggestions supply this behaviour?

**Comment by fujifish** [comment]
Busmq was created with a specific need in mind - to reliably deliver a stream of **messages**. Imagine WhatsApp messages being delivered to your phone. It wouldn't be so great if messages weren't guaranteed to be displayed in the correct order...
Busmq does not claim to solve the reliable "job/event" problem, there are a plethora of other modules out there that can do that, as mentioned in the previous comments (also take a look at the [busmq intro](https://github.com/capriza/node-busmq#why-yet-another-queue-backed-by-redis-module))

To your question, obviously having multiple consumers on the same queue in unreliable mode has little meaning for guaranteed order, but there might be cases where unreliable mode is enough; think streaming a video - it's not that bad if frames get lost.

To try and understand if busmq might be a good fit for your scenario, can you please share how these events get fed into your system in the first place, and specifically into busmq?

**Comment by tlenex** [comment]
@CodeJjang none can solve all your needs. In my example I use combo of kafka and busmq. I use kafka to stream events(small data) between microservices freely. Then I have a node app based on busmq, which listens to kafka events (each event is written as JSON, so I can easily parse it and filter them from topics) and sends them as messages in same ordered way to browser client(s).

I didn't use RabbitMQ, but I strongly recommend using kafka as inter-microservice event bus. Beside order and multiple consumers support it offers you scalability in each topic by `partitions`, when traffic gets heavy on one topic this allows you to divide the traffic by number of partitions on a topic.
Be aware, it is meant to stream events, consuming large event-messages can be very slow - best way to get big data is direct call to producer's api when the right event triggers.
Kafka lacks of browser support too, which busmq has.

All alternatives in this kind of software depends on your needs, read carefully what problem they solve and how good is their performance for it.

**Comment by CodeJjang** [comment]
@fujifish I encapsulate a "job" in a message. The message tells what kind of job needs to be done, along with it's parameters. Then the consumer service uses appropriate service to handle the job type it encountered. I do not want to lose jobs though.

Specifically, I record videos from many sources, and they're "thrown" into some filesystem in a very high rate.
Then I'd like to start a chain of jobs (some are depended on each other, and some aren't, so they could be parallelized). But I do NOT want to lose jobs, I want to perform all of them.

What happens now is that once each video is dropped to filesystem and is ready, I produce a job which says "there's a new video, at /some/path" etc.
Then there's a consumer which consumes this message, and produces more mesages (which represents jobs needs to be done), to different queues.

These queues are then consumed, jobs are performed, and some of the jobs re-produce more jobs to more queues.

That's the exact scnario. That's why I need reliability as well as scalability (to scale consumers on overloaded queues).

@tlenex I specified the scneario more precisely right here, do you still think no standalone solution is good enough for me? I really want to avoid combining some technologies before I'm 100% sure about that.

RabbitMQ seems to answer my requirements though. Take a look here:
http://www.rabbitmq.com/tutorials/tutorial-two-javascript.html

I can scale several workers (rabbit's consumer) on a channel (like a queue, but I guess does not preserve FIFO order - I do not really care about that), and also work RELIABLY (there's an ACK method).

At last, I can even mark the message & channel as durable - so even if rabbit will crash, the data won't be lost.

Seems exactly what I need. What do you say?

**Comment by fujifish** [comment]
That's what I noted in my previous comments: busmq is about messages, not jobs. It definitely sounds like you need a solution that revolves around jobs and RabbitMQ is a good choice.

Before developing busmq we've also experimented with RabbitMQ. We've done some stress testing on a RabbitMQ setup with default parameters (no tweaking of any sort, simply installed and ran). We pretty quickly ended up with a RabbitMQ that choked by reaching 100% CPU. Messages stopped flowing and we had to kill the server process (BTW, Rabbit is written in Erlang). I suggest you test this very carefully to verify it meets your needs, since I've heard of others who encountered problems like this with Rabbit.

I'm always open to new features and suggestions, so if you have a suggestion for how to implement a "job queue" mode, I'd be more than happy to collaborate with you on it!

**Comment by noamkatzir** [comment]
Have you tried apache storm?
http://storm.apache.org/releases/current/Rationale.html

**Comment by CodeJjang** [comment]
@fujifish could you please publish your results regarding the benchmarking of rmq?
perhaps you tested some old version of it? because it seems they've been improving alot in version 2.8.0: https://www.rabbitmq.com/blog/2012/04/17/rabbitmq-performance-measurements-part-1/

@noamkatzir I read the link, how's it different / complement  queues?
